File: java/MaximumProductSubarray.java
public class MaximumProductSubarray {
    public int maxProduct(int[] nums) {
        int answer = Math.max(nums[0], nums[nums.length - 1]);
        int current_positive = nums[0];
        int current = nums[0];
        int current_back_positive = nums[nums.length - 1];
        int current_back = nums[nums.length - 1];

        if(nums[0] == 0) {
            current = 1;
        }
        if(nums[0] < 0) {
            current_positive = 1;
        }
        if(nums[nums.length - 1] == 0) {
            current_back = 1;
        }
        if(nums[nums.length - 1] < 0) {
            current_back_positive = 1;
        }

        for(int i = 1; i < nums.length; i++) {
            current = current * nums[i];
            current_positive = current_positive * nums[i];
            current_back = current_back * nums[nums.length - i - 1];
            current_back_positive = current_back_positive * nums[nums.length - i - 1];
            answer = Math.max(answer, Math.max(Math.max(current, current_positive), Math.max(current_back_positive, current_back)));
            if(current_positive < 0) {
                current_positive = 1;
            }
            if(current == 0) {
                current = 1;
            }
            if(current_back == 0) {
                current_back = 1;
            }
            if(current_back_positive < 0) {
                current_back_positive = 1;
            }
        }

        return answer;
    }

    public static void main(String[] args) {
        MaximumProductSubarray mps = new MaximumProductSubarray();
        System.out.println(mps.maxProduct(new int[]{0, 2}));
        System.out.println(mps.maxProduct(new int[]{2, 2, 0, 2, 2, -2, -2, -2}));
        System.out.println(mps.maxProduct(new int[]{-2, 0, -1}));
        System.out.println(mps.maxProduct(new int[]{-2, 0, -1, 2, 3, -2, 4, 5}));
        System.out.println(mps.maxProduct(new int[]{2, -5, -2, -4, 3}));
        System.out.println(mps.maxProduct(new int[]{-1, -1, 0}));
        System.out.println(mps.maxProduct(new int[]{-1, -2, -3, 0}));
        System.out.println(mps.maxProduct(new int[]{-1, 0, -2, 2}));
    }
}

File: python/maximum_product_subarray.py
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        minus_values = [1 for _ in nums]
        answer = [0 for _ in nums]

        minus_values[0] = nums[0] if nums[0] < 0 else 0
        answer[0] = nums[0]
        maximum = answer[0]
        for i in range(1, len(nums)):
            answer[i] = max(answer[i - 1] * nums[i], minus_values[i - 1] * nums[i], nums[i])
            maximum = max(maximum, answer[i])
            minus_values[i] = min(answer[i - 1] * nums[i], minus_values[i - 1] * nums[i], nums[i])

        return maximum



s = Solution()
print(s.maxProduct([1,2,-1,-2,2,1,-2,1,4,-5,4]))
print(s.maxProduct([2,3,-2,4]))

print(s.maxProduct([2,3,-2,-1, 0, 1, 2, 100]))
print(s.maxProduct([-2,0,-1]))
print(s.maxProduct([2, -5, -2, -4, 3]))
print(s.maxProduct([7, -2, -4]))
print(s.maxProduct([-2]))
print(s.maxProduct([2,-1,1,1]))